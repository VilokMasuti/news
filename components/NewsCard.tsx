'use client';

import { formatDate } from '@/lib/utils';
import type { NewsArticle } from '@/types';
import Image from 'next/image';
import Link from 'next/link';
import { useState } from 'react';

interface NewsCardProps {
  article: NewsArticle;
  index: number;
}

export default function NewsCard({ article, index }: NewsCardProps) {
  const [imageLoaded, setImageLoaded] = useState(false);

  // Create a URL-friendly slug from the title
  const createSlug = (title: string) => {
    return title
      .toLowerCase()
      .replace(/[^\w\s-]/g, '')
      .replace(/\s+/g, '-')
      .replace(/-+/g, '-');
  };

  // Calculate animation delay based on index (for staggered entry)
  const getAnimationDelay = () => {
    return `${index * 0.09}s`;
  };

  return (
    <div
      className="group h-full flex flex-col card-hover animate-scale-in"
      style={{ animationDelay: getAnimationDelay() }}
    >
      <Link
        href={`/article/${createSlug(article.title)}?url=${encodeURIComponent(
          article.url
        )}`}
        className="flex flex-col h-full overflow-hidden rounded-lg border bg-card text-card-foreground transition-all hover:shadow-md"
      >
        <div className="relative aspect-[16/9] overflow-hidden">
          <div
            className={`absolute inset-0 bg-muted animate-pulse ${
              imageLoaded ? 'hidden' : 'block'
            }`}
          />
          {article.urlToImage ? (
            <Image
              src={article.urlToImage}
              alt={article.title}
              fill
              sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
              className={`object-cover transition-transform duration-500 group-hover:scale-105 ${
                imageLoaded ? 'opacity-100' : 'opacity-0'
              }`}
              onLoad={() => setImageLoaded(true)}
            />
          ) : (
            <div className="flex h-full items-center justify-center bg-muted">
              <span className="text-sm text-muted-foreground">
                No image available
              </span>
            </div>
          )}
          {article.source.name && (
            <div className="absolute top-3 left-3">
              <span className="px-3 py-1 text-xs font-medium rounded-full bg-secondary text-secondary-foreground transition-colors bg-amber-200">
                {article.source.name}
              </span>
            </div>
          )}
        </div>

        <div className="flex flex-col flex-grow p-4">
          <h3 className="font-semibold text-lg tracking-tight mb-2 line-clamp-2 group-hover:text-primary transition-colors">
            {article.title}
          </h3>
          <p className="text-muted-foreground text-sm line-clamp-3 mb-4">
            {article.description || 'Read the full article for more details.'}
          </p>
          <div className="mt-auto pt-4 flex items-center justify-between text-xs text-muted-foreground border-t">
            <time dateTime={article.publishedAt}>
              {formatDate(article.publishedAt)}
            </time>
            <span>{article.author || article.source.name}</span>
          </div>
        </div>
      </Link>
    </div>
  );
}
