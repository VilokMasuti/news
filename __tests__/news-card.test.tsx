/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-explicit-any */
import NewsCard from '@/components/NewsCard';
import { render } from '@testing-library/react';

// Improved Next.js Image mock
jest.mock('next/image', () => ({
  __esModule: true,
  default: (props: any) => {
    // Filter out Next.js specific props
    const { fill, ...rest } = props;
    return <img {...rest} />;
  },
}));

describe('NewsCard', () => {
  it('renders without crashing', () => {
    const { container } = render(
      <NewsCard
        article={{
          title: 'Test Article',
          url: '#',
          urlToImage: 'https://example.com/test.jpg',
          publishedAt: '2024-01-01',
          source: {
            name: 'Test Source',
            id: null
          },
          description: 'Test description',
          author: 'Test Author',
          content: 'Test content',
        }} index={0}      />
    );

    expect(container).toMatchSnapshot();
  });
});
