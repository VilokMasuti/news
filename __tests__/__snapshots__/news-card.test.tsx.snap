// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsCard renders without crashing 1`] = `
<div>
  <div
    class="group h-full flex flex-col card-hover animate-scale-in"
    style="animation-delay: 0s;"
  >
    <a
      class="flex flex-col h-full overflow-hidden rounded-lg border bg-card text-card-foreground transition-all hover:shadow-md"
      href="/article/test-article?url=%23"
    >
      <div
        class="relative aspect-[16/9] overflow-hidden"
      >
        <div
          class="absolute inset-0 bg-muted animate-pulse block"
        />
        <img
          alt="Test Article"
          class="object-cover transition-transform duration-500 group-hover:scale-105 opacity-0"
          sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
          src="https://example.com/test.jpg"
        />
        <div
          class="absolute top-3 left-3"
        >
          <span
            class="px-3 py-1 text-xs font-medium rounded-full bg-secondary text-secondary-foreground transition-colors"
          >
            Test Source
          </span>
        </div>
      </div>
      <div
        class="flex flex-col flex-grow p-4"
      >
        <h3
          class="font-semibold text-lg tracking-tight mb-2 line-clamp-2 group-hover:text-primary transition-colors"
        >
          Test Article
        </h3>
        <p
          class="text-muted-foreground text-sm line-clamp-3 mb-4"
        >
          Test description
        </p>
        <div
          class="mt-auto pt-4 flex items-center justify-between text-xs text-muted-foreground border-t"
        >
          <time
            datetime="2024-01-01"
          >
            January 1, 2024 at 05:30 AM
          </time>
          <span>
            Test Author
          </span>
        </div>
      </div>
    </a>
  </div>
</div>
`;
